packages/images/src/main.ts

```diff
@@ -0,0 +1,131 @@
+import type { Logger } from '@netlify/dev-utils'
+import { createIPX, createIPXWebServer, ipxFSStorage, ipxHttpStorage } from 'ipx'
+
+interface ImagesConfig {
+  remote_images: string[]
+}
+
+interface ImageHandlerOptions {
+  imagesConfig?: ImagesConfig
+  logger: Logger
+  originServerAddress?: string
+}
+
+export interface ImageMatch {
+  handle: () => Promise<Response>
+}
+
+const IMAGE_CDN_ENDPOINTS = ['/.netlify/images', '/.netlify/images/']
+
+export class ImageHandler {
+  #allowedRemoteUrlPatterns: RegExp[]
+  #logger: Logger
+  #originServerURL?: URL
+
+  constructor({ logger, imagesConfig, originServerAddress }: ImageHandlerOptions) {
+    this.#logger = logger
+    this.#allowedRemoteUrlPatterns = (imagesConfig?.remote_images ?? []).reduce<RegExp[]>((acc, stringPattern) => {
+      try {
+        acc.push(new RegExp(stringPattern))
+      } catch (maybeError) {
+        const error = maybeError instanceof Error ? maybeError : new Error(String(maybeError))
+        this.#logger.warn(`Malformed remote image pattern: "${stringPattern}": ${error.message}. Skipping it.`)
+      }
+      return acc
+    }, [])
+    this.#originServerURL = originServerAddress ? new URL(originServerAddress) : undefined
+  }
+
+  private generateIPXRequestURL(imageURL: URL, netlifyImageCdnParams: URLSearchParams): URL {
+    const ipxParams: string[] = []
+
+    const width = netlifyImageCdnParams.get('w') ?? netlifyImageCdnParams.get('width')
+    const height = netlifyImageCdnParams.get('h') ?? netlifyImageCdnParams.get('height')
+
+    if (width && height) {
+      ipxParams.push(`resize_${width}x${height}`)
+    } else if (width) {
+      ipxParams.push(`width_${width}`)
+    } else if (height) {
+      ipxParams.push(`height_${height}`)
+    }
+
+    const quality = netlifyImageCdnParams.get('q') ?? netlifyImageCdnParams.get('quality')
+    if (quality) {
+      ipxParams.push(`quality_${quality}`)
+    }
+    const format = netlifyImageCdnParams.get('fm')
+    if (format) {
+      ipxParams.push(`format_${format}`)
+    }
+
+    const fit = netlifyImageCdnParams.get('fit')
+    if (fit) {
+      ipxParams.push(`fit_${fit === 'contain' ? 'inside' : fit}`)
+    }
+
+    const position = netlifyImageCdnParams.get('position')
+    if (position) {
+      ipxParams.push(`position_${position}`)
+    }
+
+    const ipxModifiers = ipxParams.join(',')
+
+    return new URL(`/${ipxModifiers || `_`}/${encodeURIComponent(imageURL.href)}`, imageURL.origin)
+  }
+
+  match(request: Request): ImageMatch | undefined {
+    const url = new URL(request.url)
+
+    if (IMAGE_CDN_ENDPOINTS.includes(url.pathname)) {
```

Comment:
nit: If you invert the check here and early-return we can have one less level of nesting, which would be helpful here.